Enhancer calling script in cell lines
```{r}
#Loading necessary libraries and functions
source("~/MM_thesis_repo/libraries_thesis.R")
source("~/MM_thesis_repo/functions_MM.R")
```


```{r}
#Creating a list of chromosomes to later filter out the odd chromosomes annotated by Homer. 
chr_list <- c(paste0("chr", seq(from=1, to=22)), "chrX", "chrY")
```


```{r}
# Set a flag to control saving outputs
save_outputs <- FALSE  # Change to FALSE if you don't want to save outputs, TRUE if outputs are to be saved 
```

```{r}
#Defining variables, reading the files 
TTseq_data <- read_tsv("~/MM_thesis_repo/general_data/T1_TTseq_rpkm.txt")
gene_list <- read_tsv("~/MM_thesis_repo/general_data/hg38_refseq_genelist.txt", col_names = TRUE)
H3K27ac_data <- read_tsv("~/MM_thesis_repo/KMS12/KMS12_H3K27ac.txt", skip = 37)
ATAC_data <- read_tsv("~/MM_thesis_repo/KMS12/KMS12_ATAC_peaks.bed", col_names = FALSE)
cell_line <- "KMS12" #specify cell line of interest 
  
#output files after Homer annotation
enhancers_annot <- read_tsv("~/MM_thesis_repo/new_data/KMS12/KMS12_enhancers_H3k27ac_annot.txt") %>%
  select(1:4, 20) %>%
  set_names(c("EnhancerID", "chr", "start", "end", "H3K27ac_tag"))
ATAC_bins_ATAC <- read_tsv("~/MM_thesis_repo/new_data/KMS12/KMS12_ATAC_ATAC_tag_bins.txt") 
ATAC_bins_H3K27ac <- read_tsv("~/MM_thesis_repo/new_data/KMS12/KMS12_ATAC_H3K27ac_tag_bins.txt")
```

```{r}
#Setting up a list for promoter coordinates which should be excluded from the enhancer list 
gene_list_promoters <- gene_list %>%
  filter(chrom %in% chr_list) %>%
  select(c(chrom, strand, txStart, txEnd, name2)) %>%
  mutate(txStart = ifelse(strand == "-", txEnd, txStart)) %>%
  select(c(chrom, txStart, name2)) %>%
  unique() %>%
  set_names("chr", "start", "Geneid") %>%
  mutate(end = start +1) %>%
  select(3,1,2,4) %>%
  mutate(promoter.start = start - 2000) %>%
  mutate(promoter.end = start + 2000)

promoters_coordinates <- gene_list_promoters %>%
  select(Geneid, chr, promoter.start, promoter.end) %>%
  set_names(c("Promoter", "chr", "start", "end"))
```

```{r}
#Choosing genes which are expressed in the cell line 
TTseq_data <- TTseq_data %>%
  select(c(Geneid, length, cell_line)) %>%
  filter(!!sym(cell_line) > 1)

#Setting up the list of expressed genes from genelist 
expressed_genes <- gene_list %>%
  select(c(chrom, txStart, name2)) %>%
  set_names("chr", "start", "Geneid") %>%
  unique() %>%
  mutate(end = start +1) %>%
  select(3,1,2,4) %>%
  filter(Geneid %in% TTseq_data$Geneid)
```

Setting up acetylation data 
```{r}
H3K27ac_data <- H3K27ac_data %>%
  select(c(1:4)) %>%
  filter(chr %in% chr_list) %>%
  set_names(c("PeakID", "chr", "start", "end")) %>%
  GRanges() %>% reduce(min.gapwidth=5000) %>%
  as.data.frame() %>%
  mutate(PeakID = paste0("PeakID", row.names(.))) %>%
  select(c(-width, -strand)) %>%
  set_names(c("chr", "start", "end", "PeakID"))
```

Excluding promoter coordinates from acetylation data  - this should be annotated to make a boxplot 
```{r}
enhancers <- setdiff(GRanges(H3K27ac_data), GRanges(promoters_coordinates), ignore.strand = TRUE) %>%
  as.data.frame() %>%
  mutate(EnhancerID = paste0("EnhancerID", row.names(.))) %>% 
  filter(width > 250) %>%
  select(6,1:3) %>%
  set_names(c("EnhancerID", "chr", "start", "end"))

if (save_outputs) {
  write_tsv(enhancers, paste0("~/MM_thesis_repo/new", cell_line, "_enhancers_for_annot.txt"))
}
```

Assigning enhancer to the nearest expressed gene, calculating the distance between them - this is needed for gene expression figures
```{r}
enhancer_gene_overlap <- nearest(GRanges(enhancers), GRanges(expressed_genes)) %>%
  as.data.frame() %>%
  set_names(c("gene_index")) %>%
  cbind(enhancers) %>%
  left_join(expressed_genes %>%
              mutate(gene_index = as.integer(row.names(.))) %>%
              select(-chr, -end) %>%
              set_names(c("Geneid", "TSS", "gene_index")),
            by="gene_index") %>%
  select(-gene_index) 
```


Find overlap between the acetylation and ATAC peaks
```{r}
ATAC_data <- ATAC_data %>%
  set_names("chr", "start", "end", "PeakID")

overlap <- findOverlaps(query=GRanges(ATAC_data), subject=GRanges(enhancers)) %>%
         as.data.frame()

overlap_ATAC_enhancers <- ATAC_data %>%
  mutate(queryHits = as.numeric(row.names(.))) %>%
  full_join(overlap, by="queryHits") %>%
  left_join(enhancers %>%
              mutate(subjectHits = as.numeric(row.names(.))) %>%
              select(c("EnhancerID", "subjectHits")), 
            by = "subjectHits") %>%
    select(-subjectHits, -queryHits) 

overlap_2 <- findOverlaps(query = GRanges(overlap_ATAC_enhancers), subject = GRanges(promoters_coordinates)) %>%
  as.data.frame()

#This data should be annotated with ATAC and H3K27ac tags for boxplot and metaplots 
ATAC_data_final <- overlap_ATAC_enhancers %>%
  mutate(queryHits = as.numeric(row.names(.))) %>%
  full_join(overlap_2, by = "queryHits") %>%
  left_join(promoters_coordinates %>%
              mutate(subjectHits = as.numeric(row.names(.))) %>%
              select(c("Promoter", "subjectHits")),
            by = "subjectHits") %>%
  select(-subjectHits, -queryHits) %>%
  mutate(peak_type = case_when(
    is.na(EnhancerID) & is.na(Promoter) ~ "other ATAC peak", 
    is.na(EnhancerID) & !is.na(Promoter) ~ "promoter", 
    !is.na(EnhancerID) & is.na(Promoter) ~ "enhancer", 
    !is.na(EnhancerID) & !is.na(Promoter) ~ "enhancer"
  ))
```

Joining the annotated enhancers tag with the ATAC peaks 
```{r}
enhancer_annot <- enhancers_annot %>%
  left_join(ATAC_data_final %>%
              select(PeakID, EnhancerID), by = "EnhancerID") %>%
  mutate(peak_size = (end - start)) %>%
  mutate(RPKM = H3K27ac_tag*1000 / peak_size) %>%
  mutate(ATAC_peak = ifelse(is.na(PeakID), "no ATAC peak", "ATAC peak")) %>%
  mutate(cell_line = cell_line) %>%
    unique() %>%
  left_join(enhancer_gene_overlap %>% 
              select(Geneid, TSS, EnhancerID), by = "EnhancerID")


write_tsv(enhancer_annot, paste0("~/MM_thesis_repo/new_data/", cell_line, "_enhancer_acetylation.txt"))
```

Making a boxplot showing acetylation reads at enhancers with or without ATAC peak 
```{r}
boxplot_enhancers <- enhancer_annot %>%
    ggplot(aes(x = ATAC_peak, y = log2(RPKM))) +
    geom_boxplot() +
  scale_color_manual(
    values=wes_palette(2, name="Darjeeling1")) +
  labs(
    x = NULL,
    y = expression(paste("log"[2], "(Number of reads)"))) +
  scale_x_discrete(labels = c("ATAC" = "with ATAC peak", 
                              "no ATAC" = "without ATAC peak")) +
  theme(
    axis.title.x = element_text(size = 12),
    axis.title.y = element_text(size = 11),
    axis.text.y = element_text(size = 10, color = "black"),
    axis.text.x = element_text(hjust = 0.5, vjust = 0.5, size = 10, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "white"),
    axis.line = element_line(size = 0.5, linetype = "solid",
                                   colour = "black"),
    legend.key.size = unit(0.8, "cm"),
  )
boxplot_enhancers
```

Making a metaplot of the acetylation at ATAC peaks 
```{r}
ATAC_hist <- ATAC_bins_ATAC %>%
  mutate(tag_type = "ATAC") %>%
  rbind(ATAC_bins_H3K27ac %>%
          mutate(tag_type = "H3K27ac")) %>%
  left_join(ATAC_data_final, by = c("Gene" = "PeakID")) %>%
  filter(!is.na(Gene)) %>%
  select(-Gene, -chr, -start, -end, -EnhancerID, -Promoter) %>%
  mutate(across(-tag_type:-peak_type, ~ log2(.x+0.1))) %>%
  group_by(tag_type, peak_type) %>%
  summarise_all(funs(mean)) %>%
  mutate(Stat = "mean")

ATAC_hist_sem <- ATAC_bins_ATAC %>%
  mutate(tag_type = "ATAC") %>%
  rbind(ATAC_bins_H3K27ac %>%
          mutate(tag_type = "H3K27ac")) %>%
  left_join(ATAC_data_final, by = c("Gene" = "PeakID")) %>%
  filter(!is.na(Gene)) %>%
  select(-Gene, -chr, -start, -end, -EnhancerID, -Promoter) %>%
  mutate(across(-tag_type:-peak_type, ~ log2(.x+0.1))) %>%
  group_by(tag_type, peak_type) %>%
  summarise_all(funs(std.error)) %>%
  mutate(Stat = "sem") %>%
  filter(!is.na(peak_type))

ATAC_hist_combined_sem <- rbind(ATAC_hist, ATAC_hist_sem) %>%
  melt(id.vars = c("peak_type", "tag_type", "Stat"), 
       variable.name = "distance_to_centre", value.name = "tag")

ATAC_hist_combined_sem$distance_to_centre <- as.numeric(as.character(ATAC_hist_combined_sem$distance_to_centre))
ATAC_hist_combined_sem$tag <- as.numeric(as.character(ATAC_hist_combined_sem$tag))

ATAC_hist_stats <- ATAC_hist %>%
  melt(id.vars = c("tag_type", "peak_type"), 
       variable.name = "distance_to_centre", value.name = "mean") %>%
  full_join(ATAC_hist_sem %>%
              melt(id.vars = c("tag_type", "peak_type"), 
                   variable.name = "distance_to_centre", value.name = "sem"),
            by = c("tag_type", "peak_type", "distance_to_centre")) 


ATAC_hist_stats$distance_to_centre = as.numeric(as.character(ATAC_hist_stats$distance_to_centre))
ATAC_hist_stats$mean = as.numeric(as.character(ATAC_hist_stats$mean))
ATAC_hist_stats$sem = as.numeric(as.character(ATAC_hist_stats$sem))

ATAC_hist_stats <- ATAC_hist_stats %>%
  mutate(ymin = mean -sem,
         ymax = mean +sem)

ATAC_hist_stats <- ATAC_hist_stats %>%
  filter(!is.na(mean)) %>%
  filter(!is.na(peak_type))
```


```{r}
legend_labels <- c("ATAC" = "ATAC reads", "H3K27ac" = "H3K27ac reads")

plt_linegraph <- ATAC_hist_stats %>%
  ggplot(aes(x = distance_to_centre, y = mean, color = tag_type)) +
  facet_grid(.~peak_type) +
  geom_line() +
    scale_x_continuous(breaks = c(-750, 0 , 750), labels = c("-750", "0", "750"))
  scale_color_manual(
    values=wes_palette(2, name="GrandBudapest1"), labels = legend_labels, name = "Tag used for annotation") +
  labs(
    x = "Distance from peak centre (bp)",
    y = expression(paste("log"[2], "(Number of reads)"))) +
  theme(
    axis.title.x = element_text(size = 12),
    axis.title.y = element_text(size = 12),
    axis.text.y = element_text(size = 10, color = "black"),
    axis.text.x = element_text(size = 10, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "white"),
    axis.line = element_line(size = 0.5, linetype = "solid",
                                   colour = "black"),
    legend.key.size = unit(0.8, "cm"),
    legend.title = element_text(size=10),
    strip.text.x = element_text(size=10),
    panel.spacing = unit(1.5, "lines"),
    strip.background = element_rect(colour="black", fill="white", 
                                       size=0.8)
  ) 
plt_linegraph

if (save_outputs) {
  ggsave((paste0("~/myeloma_thesis_R/new_data/figures/", cell_line, "_ATAC_acetylation_metaplot.pdf")), plt_linegraph)
  ggsave((paste0("~/myeloma_thesis_R/new_data/figures/", cell_line, "_ATAC_acetylation_metaplot.jpg")), width = 7, height = 5, plt_linegraph)
}
```

